FK_solver/src/FK_solver.hpp

```diff
@@ -0,0 +1,282 @@
+// #ifndef HEAT_NON_LINEAR_HPP
+// #define HEAT_NON_LINEAR_HPP
+#ifndef FK_SOLVER_HPP
+#define FK_SOLVER_HPP
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_values_extractors.h>
+#include <deal.II/fe/mapping_fe.h>
+
+#include <deal.II/grid/grid_in.h>
+
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <fstream>
+#include <iostream>
+
+using namespace dealii;
+
+// Class representing the non-linear diffusion problem.
+class FK_solver
+{
+public:
+  // Physical dimension (1D, 2D, 3D)
+  static constexpr unsigned int dim = 3;
+
+
+  // Function for the alpha coefficient.
+  class FunctionAlpha : public Function<dim>
+  {
+  public:
+    virtual double
+    value(const Point<dim> & /*p*/,
+          const unsigned int /*component*/ = 0) const override
+    {
+      return 0.9;
+    }
+  };
+
+  // Function for the forcing term.
+  class ForcingTerm : public Function<dim>
+  {
+  public:
+    virtual double
+    value(const Point<dim> & /*p*/,
+          const unsigned int /*component*/ = 0) const override
+    {
+      return 0.0;
+    }
+  };
+
+  // Function of the matrix D
+  class FunctionD : public Function<dim>
+  {
+  public:
+    virtual void
+    matrix_value(const Point<dim> & /*p*/,
+                 FullMatrix<double> &values) const // override
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        {
+          for (unsigned int j = 0; j < dim; ++j)
+            {
+              if (i == j)
+                values[i][j] = 0.1;
+              else
+                values[i][j] = 0.0;
+            }
+        }
+        values[1][1]+=10.0;
+    }
+    virtual double
+    value(const Point<dim> & /*p*/,
+          const unsigned int component_1 = 0,
+          const unsigned int component_2 = 1) const // override
+    {
+      if (component_1 == component_2)
+        return 1.0;
+      else
+        return 0.0;
+    }
```

Comment:
Here you are changing the interface of the class `dealii::Function` from which you are inheriting.

While C++ does admit this kind of thing, it is generally not a good idea from a design perspective, because it suggests that objects of type `FunctionD` can be used whenever an object of type `Function` is needed - which is not true.

To make an example, you could pass an object of type `FunctionD` to the function `dealii::VectorTools::interpolate` and everyhing would compile just fine. However, `interpolate` would try to call `FunctionD::value(const Point<dim>&, const unsigned int)` (notice the single `unsigned int` argument). This method is defined in the base class `Function`, and it is never overridden in the derived class. Hence, an user might think they were interpolating the tensor D, but in reality they are interpolating whatever the default implementation of `Function::value` contains. In medium-sized projects, this kind of thing will inevitably lead to hard-to-find bugs.

tl;dr: if you don't intend to use the interface of a certain class (as in this case), you should not inherit from that class.